### DISCUSSION TITLE
Catalog Repository Structure

### DISCUSSION THREAD

**Discussion by dave-kaye** [XEphem/XEphem#39] [Ideas]
This discussion thread is being placed for response to an idea for modification of the layout used to display the XEphem catalogs.

The present structure is for existing catalogs.  A slight modification in structure will serve for ongoing catalog release and user interaction.  The proposed structure is:

-------------------------------------------------
#  XEphem Catalogs

The star catalogs in this repository are for the
XEphem serious interactive astronomical software ephemeris.

## Original

* **`GSC2201/`**

* **`usno/`**

* **`gsc/`**

* **`Official/`**

## User contributed
-------------------------------------------------

The present master header of Official would be replaced with Original, and a description of the catalog background taken from the official directory link could be used.  This layout more accurately describes the available catalogs as a group.  The three major field star catalogs which were released as directories would be broken out to top headers where a description of the catalog and purpose could be given.  Users would be able to see and select those they want.  The usno and gsc directories are presently in a pull request awaiting disposition, and are located under the official directory in that request.  

The Official directory would keep the star catalogs, with the exception of the download catalogs from Celestrak and Minor Planet Center Soft03 collection.  These are orbital satellites, asteroids and comets.  They have a life span of about one-half year, and are presently a decade old.  These would be deleted and the existing directory readme file updated (about 5 line change).  The major catalogs from Lowell and Minor Planet Center accessed with a Perl script could also be removed with a larger impact on the readme file, as they comprise their own separate section.

A new master header of Download would be created with an accompanying directory.  This would be a snapshot in time of the data -> download catalogs.  How-to description of the download method for up-to-date catalogs would be included here. This will allow people access to the catalogs where they may have problems with the site connections.  The download catalogs will number about 10, and can be updated every 4-6 months in a matter of a few minutes.  A XEphem date header internal to these catalogs has been proposed in a code pull request that is awaiting disposition. It would make it easier for the user to know the age of these catalogs.

A master header of Update would hold updated catalogs from the original XEphem distribution.  As these would be contributed by users, I feel they should be separated from Downey's extensively checked releases and not simply overwrite the catalogs in the Official directory.  As they are user contributions, this section could be merged into the user contributed section and not be called out as a master header.

A master header of User contributed would remain for ongoing catalog submissions.  If this should get to large in the future, it can be dealt with then.

-------------------------------------------------

While the layout would change, I propose a change in function as well.  If you want the catalogs, you can only get them if you have GIT access or by using the web page code button creating a zip file of the entire system.  This will not be easy for a large number of users.  It also is not obvious to users who are not acquainted with GitHub.  A download link to download each individual directory in the original section should be provided.  The download/update/user contributed sections could be downloaded as the whole directory with separate download links.

GitHub is designed not to do this, and much discussion on the internet focusing on downloading individual directories exists.  There are workarounds that people have prepared, but many of them are not practical in this situation.  I do have a fork of the catalog repository where I am experimenting with several options.  It _may be_ possible to provide these links using a combination of techniques.

-------------------------------------------------

I am not in charge of the area, so this is not an official direction in the future.  This is simply a concept suggestion, and comments are invited.

**Comment by brandon-rhodes** [comment]
Thanks for thinking through how the catalog README can evolve — I'm still reading over your suggestions and thinking about them.

But you point out one glaring problem that we can get to work on immediately: the difficultly of downloading the big catalogs! I think there's a simple solution: while GitHub doesn’t support downloading a subdirectory of a repository, it does support downloading a branch. So for each big catalog we can create a new branch, delete all the files there except for one big catalog, and then use that branch to offer a URL to users. And it will be very storage efficient, since the files themselves will only exist once each on GitHub and in the git repository.

Let's try. I've just created a sample branch, and you can grab its zip file here:

https://github.com/XEphem/Catalogs/archive/refs/heads/GSC2201.zip

It's unfortunate that the zip file includes a redundant directory. The zipfile contents look like:

```
Archive:  /home/brandon/Downloads/Catalogs-GSC2201.zip
6a2ffea09c9f98e5289aa6a4c380ff1e4d7bd67c
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2021-11-30 18:01   Catalogs-GSC2201/
        0  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/
  7724303  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/N00.xe3
  7674689  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/N01.xe3
  8574605  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/N02.xe3
  8272496  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/N03.xe3
...
   814850  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/S86.xe3
   515648  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/S87.xe3
   185558  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/S88.xe3
    57647  2021-11-30 18:01   Catalogs-GSC2201/GSC2201/S89.xe3
---------                     -------
1355418339                     182 files
```

So folks will have to unzip, then grab the inner directory `GSC2201` and put that — not the outer directory — into a path that XEphem can find.

What do you think? Would a branch-per-big-catalog like this solve the no-zipfile problem? Or is it too unwieldy?

**Reply by dave-kaye** [reply]
Test download failed.  It took a very long time to download the file from the web.  When I was able to compare the resultant catalog with the original catalog, the last 25% sections were missing (starting with S44.xe3).  When I retried to unzip the file, the first sections were missing and only the remaining 25% was there.  There is something in the download at that point that is an issue.  I will try to download and test again.

This does bring up a good point however.  If a large catalog is being downloaded as a single file, probably an MD5 or similar value should be listed on the website so the user can check to see if the download was good.  This will only work if the catalog file is stored as a single file, and not being dynamically created by the website as it is downloaded.

**Reply by brandon-rhodes** [reply]
I am perplexed. I thought zip files put their directory of contents at the end of the file? So that if a download wasn't complete, then the zipfile wouldn't be useable?

Could you check whether your disk is full?

**Reply by dave-kaye** [reply]
Tried second complete download and then unzip process.  Download much faster than first time.  All sections were present and were identical to the original.

I have had multiple occasions where files were missing from a dynamic creation of zip, the system just never added them to the zip and closed it out.

Here there might be a zip issue on the first try if dynamic, otherwise a download error of a static file with a couple of bytes bad would produce an unusual  file.  No, I don't understand what happened either, it just did.  No problems on PC or unzip ends.

**Comment by dave-kaye** [comment]
Downloading test link now.  Will have to take home and test if it matches the original with group diff test.

There may be a different way.  See github "dreamchess" webpage.  They have a series of files called assets that could be a way to encapsulate the single directory.  Also, you may be able to do a release of a partial of the website, I was going to look into GitHub capability on that.  Finally, you could just create a readme.txt file, pull the directory to top level, delete all the remaining files and add the readme:

readme.txt - what it is and what to do with it for XEphem
gsc2200/

Release it and call it GSC2200.zip and it is listed in the release column.  Easy url for download then and can be put on the webpage as url.  Directory structure is now intact in the zip file I think.  It always seems to be intact for main code downloads.  But I think you must do this on the main branch, I don't think you can release from a side branch.

Now reload the website and repeat.  Would only need to do 3 total single databases and probably one for the multi-listing original directory.  These could each keep a different zip name to match the catalog inside.  These are historical, they will never change so the release will never need to be updated.  (If doing the multi, contact me - we should pull out some junk and update the directory readme).  Another thought - it may not be necessary to reload all the catalogs at the end of the process.  They would exist in the release and not be needed in the active code area anymore.  Just need to keep the archives of them viable.

For the small stuff, I tested one download system and there is a second to look at.  Downgit works well on the small stuff.  You can access the page by looking at my catalog fork, then the branch dnload.  I am keeping a running commentary there to tell of the results.  The second method would not use an external website, it uses some code directly on the site (I think).  The program is gitzip.  There are others, but they require downloading files to put into your browser on a permanent basis.  Not feasible.

Why downgit?  Because if the catalog structure remains as is, then the only way to get the small stuff would be by downloading the whole catalog directory of GB size - back to the original problem.  And these would be dynamic so release technique would have to keep on releasing for a maintenance headache.  You could zip and put in subdirectory (see test page) but once again you would have to zip on every change.

There may be a way to get a url to download a single catalog (User Contributed).  But a lot of times these just show up in the browser and not as a download.  Could look at later for feasibility.

Back to your test.  I would put a readme.txt file external to the directory at top level so people would know what to do.  Branches may work, but you are going to get a lot of branches showing up, and still have the directory issue.  But as a solution, it is far better than the present way.

**Comment by dave-kaye** [comment]
If the Official directory keeps the star catalogs with the exception of removing the download catalogs from Celestrak and Minor Planet Center Soft03 collection due to age, then the only change to the directory ReadMe.txt would be the removal of the line for catalog comets.edb (approximately line 208). This catalog is not in the directory, I am unable to find it, and it appears to be 2 decades old.  The files to be removed that are from Celestrak and Minor Planet Center Soft03 collection (science.edb, stations.edb, Soft03Cmt.edb, Soft03CritList.edb, Soft03Distant.edb, Soft03Unususal.edb) are not listed in the readme.  Repeating an earlier post, these files are decades old, are good for only one year, and can be freshly prepared via the download section of the data menu of XEphem.

These changes would bring the ReadMe.txt listing and actual files present into alignment.

Another old catalog is spacecraft.edb which is decades old and in the readme file.  Its data structure is unique, and does not show its orgin.  But it does contain some historical satellites and should therefore be kept as is.

**Comment by dave-kaye** [comment]
I had suggested a readme.txt file in the catalog zip file. It would be at the top level so you could see it immediately, and not be in the GSC2201 directory. This is a sample of what I was thinking about.
[ReadMe.txt](https://github.com/XEphem/XEphem/files/7654666/ReadMe.txt)

**Comment by dave-kaye** [comment]
It has been some time since this discussion was raised. In revisiting it, I have an improved version for the catalog structure.

It would start with the creation of a third repository for XEphem, **_FieldStars_**. It would contain all the field star catalogs that have been released. In it would be moved the **GSC2201** directory (GSC 2.2.0.1 catalog) from the "_Catalog_" repository.  The **usno** directory (USNO-SA2.0 catalog) and the **gsc** directory (GSC-ACT catalog) which are found in pull request number 2 from the "_Catalog_" repository would also be moved into this new repository.  The **hiptyc2.xe2** file (Hipparcos/Tycho-2 catalog) and **ppm.xe2** file (Positions and Proper Motion catalog) would be moved from the "_Catalog_" repository "_Official_" directory to this new repository.

A sample readme file which will show the indicated repository structure is here:

[README.md](https://github.com/XEphem/XEphem/files/14999109/README.md)

This would put the large field star catalogs into one repository that would only have to be downloaded once. This would help alleviate the issue of the large downloads in the catalog repository.  It would take them out of the catalog repository that would be updated more frequently.  I have tried to make zip files on-the-fly for these catalogs, but have had unfavorable results. Everything I have tried/seen uses outside servers which are unreliable and a potential virus source.

The remainder of the concept would stay the same, with the "_Catalog_" repository having Official (or Original), Updates, Downloads and User Contributed sections as outlined previously.  @rmathar has a pull request in the "_Catalog_" repository for several of the outdated catalogs to be replaced.  I would place them in a Downloads section, per previous.  These may be technically user contributed catalogs, but they use the existing XEphem codes to generate a replacement catalog and are less of a user created catalog. It could also be argued that they could reside in the Updates section and not have a Downloads section. I would also at this time remove all the old catalogs from the _Official_ directory that were created in the XEphem download files section.  They were designed to be updated with changing data and not remain stagnant with out-of-date, now incorrect data.

**Comment by dave-kaye** [comment]
Would you please look at cleaning the XEphem Catalog repository.

The best option that I can think of is the creation of a _XEphem FieldStars_ repository.  This repository would take the large directories from the Catalog directory and add the field stars directories that were in XEphem 3.7 commercial version.  This would then shrink the catalog directory so it is easier to download and it would be comprised of only one style of catalog (full star data catalog).

Specifically, take the _GSC2201_ directory from the Catalog repository and move it to the top line of the FieldStars repository directory.  Take the _Official/hiptyc2.xe2_ and _Official/ppm.xe2_ files from the Catalog repository and move them to the top line of the FieldStars repository directory while simultaneously removing the Official directory level from their directory tree.

A sample readme file for this repository is:

[Readme.md](https://github.com/user-attachments/files/28808244/Readme.md)


The catalog structure is shown in this mockup image:

<img width="1306" height="800" alt="mockup field star repository page" src="https://github.com/user-attachments/assets/cf39c50f-a9b3-4db6-bc57-b84f223472de" />


The XEphem program calls for field star catalogs that are no longer readily available.  It is important to make them available here.

There will not be any major changes immediately to the FieldStars repository, so create a release 1.0 from the existing files. Name the release "Field Stars" or something similar. Github allows assets to be attached to a release - a binary file up to 2 GB and 1000 asset files maximum per release.  You simply upload them while making the release. Take the Official/gsc and Official/usno directories from the Catalog repository pull request no. 2 and remove the Official directory level from their directory tree. Take each of the directories and compress them to *.zip and/or *.tar.gz.  This will allow these larger files to be downloaded individually as the user desires.  The compressed size of USNO-SA2.0 is 452 MB and GSC-ACT is 138 MB -- well under the 2 GB maximum. Make all of the compressed and released files have an internal directory structure so they can be immediately placed in the user's XEphem catalog directory upon expansion. Name the asset files by catalog name (not directory name) since several versions of catalogs can be substituted in the directory. 

I have tried to make automatic download sequences for parts of a repository by using non-supported github projects without any success.  The use of assets will allow the repository to be broken up into individual large catalogs.

When the field star catalogs are available, I can go through the XEphem code and update the options so they make sense; as well as update the help file which is several code changes out-of-date.


**************************************

A second item is to remove the files from the catalog repository _Official/_ directory that XEphem also creates using the download function.  These are comet, asteroid and satellite catalogs that are constantly being updated. The ones in the catalog repository are years to near decades out-of-date.

The files to be removed are:

- AstLowell_dim.edb
- AstLowell.edb
- AstMPC_dim.edb
- AstMPC.edb
- science.edb
- Soft03Cmt.edb
- Soft03CritList.edb
- Soft03Distant.edb
- Soft03Unusual.edb
- spacecraft.edb
- stations.edb

Some of these files are no longer being prepared and other substitute files are being made available via the download function.  It is easier to simply remove them rather than to require a constant maintenance effort.  Otherwise, create a separate directory for the downloads and then update them every 3-6 months.
